### PR TITLE
#161291005 Build Out User Password Reset Feature

### PR DESCRIPTION
**Pull request by mrSerious**
#### What does this PR do?

Have a user password reset feature

#### Description of Task to be completed?

- add tests
- add routes
- add controller methods
- add route validation

#### How should this be manually tested?

N/A

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?

[#161291005](https://www.pivotaltracker.com/n/projects/2205593/stories/161291005)

#### Screenshots (if appropriate)

Route that will send the reset password link.
screen shot 2018-11-06 at 12 20 42 pm

Rout that will reset the password.
screen shot 2018-11-06 at 12 28 52 pm

#### Questions:
N/A

### CODE REVIEW

File: server/controllers/UsersController.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
 import bcrypt from 'bcrypt-nodejs';
+import jwt from 'jsonwebtoken';
 import models from '../models';
 import createToken from '../helpers/createToken';
 import sendEmail from '../helpers/sendEmail';
 import verifyEmailMessage from '../helpers/verifyEmailMessage';
+import resetPasswordEmail from '../helpers/resetPasswordEmail';
+import forgotPasswordEmail from '../helpers/forgotPasswordEmail';
```
Comment (andela-moe) [inline server/controllers/UsersController.js]: Seeing that both _resetPasswordEmail_ and _forgotPasswordEmail_ are basically functions serving as templates for the possible emails we need to send, you could have all possible templates in one file, each one exported. That way, you can import (destructure) both functions from the same file.

import { forgotPasswordEmail, resetPasswordEmail } from '../helpers/emailTemplates';

Then your emailTemplates.js can have something similar to what is below.
```
export const forgotPasswordEmail = resetUrl => ({
  subject: 'Authors Haven Password Reset',
  body:
    `<p>
      Click <a href="${resetUrl}">here</a> to reset your password.
    </p>`
});
```

```
export const resetPasswordEmail = () => ({
  subject: 'Authors Haven Password Reset',
  body:
    `<p>
      Your password has successfully been reset.
    </p>`
});
```

File: server/helpers/forgotPasswordEmail.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const resetPasswordEmail = resetUrl => ({
+  subject: 'Authors Haven Password Reset',
+  body:
+    `<p>
+      Click <a href="${resetUrl}">here</a> to reset your password.
+    </p>`
+});
+
+export default resetPasswordEmail;
```
Comment (andela-moe) [inline server/helpers/forgotPasswordEmail.js]: A base filename should exactly match the name of its default export. Your file name is _forgotPasswordEmail.js_ but your default export is _resetPasswordEmail_

File: server/controllers/UsersController.js (hunk 1 above)
Comment (mrSerious) [reply server/controllers/UsersController.js]: Okay, I will fix this.

File: server/helpers/forgotPasswordEmail.js (hunk 2 above)
Comment (mrSerious) [reply server/helpers/forgotPasswordEmail.js]: My mistake, I will sort this out.

File: server/helpers/emailTemplates.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+ */
+const resetPasswordEmail = () => ({
+  subject: 'Authors Haven Password Reset',
+  body:
+    `<p>
+      Your password has successfully been reset.
+    </p>`
+});
+
+export { resetPasswordEmail, verifyEmailMessage, forgotPasswordEmail };
```
Comment (mrSerious) [inline server/helpers/emailTemplates.js]: @andela-moe this is what the file looks like now.

File: server/controllers/UsersController.js [hunk 4]
```diff
@@ -98,6 +100,132 @@ class UsersController {
...
+                      message: err.message,
+                    },
+                  });
+              });
+          });
+      });
+    } else {
+      return res.status(400).send({
+        status: 'failure',
+        message: ['Passwords did not match']
```
Comment (mbilesanmi) [inline server/controllers/UsersController.js]:
```suggestion
        message: ['Passwords do not match']
```

File: server/controllers/UsersController.js [hunk 5]
```diff
@@ -98,6 +100,132 @@ class UsersController {
...
+            const hashedPassword = bcrypt.hashSync(`${password}`);
+
+            user.update({ password: hashedPassword })
+              .then(() => {
+                const message = resetPasswordEmail();
+
+                // send email
+                sendEmail(user, message);
+
+                res.status(200).send({
```
Comment (mbilesanmi) [inline server/controllers/UsersController.js]: When the user resets password successfully, I'm not sure we want to return all of his info. BTW, are we assuming that once this happens he is auto signed in? If so, are we generating an auth token for this?

File: server/middlewares/UserValidation.js [hunk 6]
```diff
@@ -22,6 +22,59 @@ class UserValidation {
...
+  /**
+    * @description - This method validates the user request body.
+    * @param {object} req - The request object to be validated.
+    * @param {object} res - Th response object to be validated.
+    * @param {object} next - The callback function to the next middleware.
+    * @returns {object} - The error object with message.
+    * @memberOf UserValidation
+    * @static
+    */
+  static validateReset(req, res, next) {
```
Comment (mbilesanmi) [inline server/middlewares/UserValidation.js]: ValidateReset what???

File: server/middlewares/UserValidation.js [hunk 7]
```diff
@@ -22,6 +22,59 @@ class UserValidation {
...
+  /**
+    * @description - This method validates the user request body.
+    * @param {object} req - The request object to be validated.
+    * @param {object} res - Th response object to be validated.
+    * @param {object} next - The callback function to the next middleware.
+    * @returns {object} - The error object with message.
+    * @memberOf UserValidation
+    * @static
+    */
+  static validateForgot(req, res, next) {
```
Comment (mbilesanmi) [inline server/middlewares/UserValidation.js]: validateForgot what??

File: server/controllers/UsersController.js (hunk 5 above)
Comment (mrSerious) [reply server/controllers/UsersController.js]: You are right about the bio, I will take that out. My idea for what happens after a password reset was for the user to get routed to the login form.

File: server/controllers/UsersController.js (hunk 4 above)
Comment (mrSerious) [reply server/controllers/UsersController.js]: I will make this alteration. Thank you.

File: server/middlewares/UserValidation.js (hunk 7 above)
Comment (mrSerious) [reply server/middlewares/UserValidation.js]: validateForgotPassword. I will make this change now.

File: server/middlewares/UserValidation.js (hunk 6 above)
Comment (mrSerious) [reply server/middlewares/UserValidation.js]: validateResetPassword, I make this alteration now.

File: server/controllers/UsersController.js [hunk 8]
```diff
@@ -98,6 +100,128 @@ class UsersController {
...
+            errors: {
+              message: ['No account with that email address exists']
+            }
+          });
+        }
+
+        const userId = userFound.id;
+        const token = createToken(userId, lifeSpan);
+        const forgotUrl = `${req.protocol}://${req.host}/password_reset`
+        + `/${token}`;
```
Comment (Darthrighteous) [inline server/controllers/UsersController.js]: This `forgotUrl` will be `/password_reset/token`. This is different from the actual route you created which is `/api/v1/users/reset/:token`. So when the user clicks the link, it won't trigger the route. 

Everything else looks good, so good job on that.

Comment (mrSerious) [reply server/controllers/UsersController.js]: Yeah, I can see now, thanks for spotting this error.

File: server/controllers/UsersController.js (hunk 5 above)
Comment (mbilesanmi) [reply server/controllers/UsersController.js]: If so, don't we need to generate a login token for them?

Comment (mrSerious) [reply server/controllers/UsersController.js]: I don't think so, the user should get a token after they have logged in.

File: server/controllers/UsersController.js [hunk 9]
```diff
@@ -101,6 +103,128 @@ class UsersController {
...
+            errors: {
+              message: ['No account with that email address exists']
+            }
+          });
+        }
+
+        const userId = userFound.id;
+        const token = createToken(userId, lifeSpan);
+        // eslint-disable-next-line max-len
+        const forgotUrl = `${req.protocol}://${req.get('host')}/api/v1/users/reset/${token}`;
```
Comment (Darthrighteous) [inline server/controllers/UsersController.js]: we are now using an env variable for the apps base url right? I think this should be updated too if we're going with that

**Pull request closed without merging**